include/swift/Basic/TypeID.h

```diff
@@ -0,0 +1,99 @@
+//===--- TypeID.h - Simple Type Identification ------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the TypeID template, which provides a numeric
+//  encoding of (static) type information for use as a simple replacement
+//  for run-time type information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_TYPEID_H
+#define SWIFT_BASIC_TYPEID_H
+
+#include <cstdint>
+#include <vector>
+
+namespace swift {
+
+/// Form a unique 64-bit integer value describing the type `T`.
+///
+/// This template needs to be specialized for every type that can
+/// participate in this kind of run-time type information, e.g., so
+/// that it can be stored in \c AnyValue.
+template<typename T>
+struct TypeID;
+
+/// Each type "zone" can contain 256 separate types for use in typeid,
+/// which should be enumerated.
+struct TypeIdZones {
+  enum ZoneValues : uint8_t {
+    /// C/C++ language and standard library types.
+    C = 0,
+
+    /// A zone used only for unit testing.
+    UnitTests = 1,
+
+    /// A zone used for the Swift type checker.
+    TypeChecker,
+  };
+};
+
+/// Form a type ID given a zone and type value.
+constexpr uint8_t formTypeID(uint8_t zone, uint8_t type) {
+  return (zone << 8) | type;
```

Comment:
Hm, this looks like overflow; should this return `uint16_t`? (and are casts needed to stop the intermediate overflowing?)